docs/src/references/mappers/c8y-mapper.md

```diff
@@ -1325,3 +1325,121 @@ te/device/main///cmd/log_upload/<cmd_id>
+
+## Message processing issues
+
+### Out of order delivery of registration and non-registration messages
+
+Since the registration messages and other messages are sent to different topics,
+even if a publishing client sends them in order: registration message followed by a twin update message,
+the MQTT broker may deliver those messages in the reverse order to the mapper,
+as MQTT protocol does not guarantee any ordering across topics.
+This case is illustrated in the following diagram:
+
+```mermaid
+sequenceDiagram
+    participant Child device
+    participant MQTT Broker
+    participant Tedge Mapper
+    
+    Child device ->> MQTT Broker: Child device registration message to `te/floor1/plc1//`
+    Child device ->> MQTT Broker: Twin update message to `te/floor1/plc1///twin/hardware_info`
+    Child device ->> MQTT Broker: Twin update message to `te/floor1/plc1///twin/location`
+    MQTT Broker ->> Tedge Mapper: Twin update message to `te/floor1/plc1///twin/location`
+    Tedge Mapper ->> Tedge Mapper: Drop the twin update message
+    MQTT Broker ->> Tedge Mapper: Twin update message to `te/floor1/plc1///twin/hardware_info`
+    Tedge Mapper ->> Tedge Mapper: Drop the twin update message
+    MQTT Broker ->> Tedge Mapper: Child device registration message to `te/floor1/plc1//`
+```
+
+When the mapper receives the twin update messages first, they are ignored as that entity is not registered yet.
+Even though the entity registration message is processed later, the ignored twin updates are not processed again,
+until the mapper is restarted, and these retained messages are re-delivered to the mapper.
+
+#### Initial solution thoughts
+
+##### Solution 1
+
+One obvious solution approach would be to cache all the non-registration messages,
+until the registration message eventually arrives, post which the cached messages are also processed.
+
+But, in-memory caching of these messages could be problematic, as they might grow unexpectedly,
+if the registration message delivery gets delayed, resulting in OOME risks.
+Writing them to the disk, to an append-only-log or something, would help avoid the OOME issue,
+but that comes with additional implementation complexity.
+
+Even with disk persistence, any indefinite waits will have to be prevented with timeouts,
+but that comes with additional uncertainty: what to do if the registration message gets delivered after the timeout?
+Crash itself (the mapper) with the hope of timely message delivery after the restart?
+
+##### Solution 2
+
+An alternate approach would be to send a feedback for entity registration messages,
+which the clients must wait on, before sending any other data associated to that entity.
+The feedback can be sent on the same topic on which the registration message was received on,
+with a field like `"@registered" : "true"`, in addition to any other enhancements done by tedge,
```

Comment:
There's a problem with this approach when there are multiple mappers processing the same registration message. Who will add that `@registered` flag? They all can't add the same flag. We can let each mapper add its own cloud-specific flags like `@c8y-registered`, `@az-registered` etc, but then, the device side logic, which needs to wait for the registration complete acknowledgement, will have to wait for all those cloud specific flags separately, pushing cloud specific logic into the device, defeating the cloud-agnosticism that we promise.

Sending these "enriched registration messages" to separate cloud specific topics is even worse, as they suffer from the same limitation of "cloud specific logic pushed to device", with the overhead of multiple message copies per cloud for the same entity.

Reply:
On a second thought, allowing all the mappers to enrich the source registration message by publishing to the same topic doesn't look good, as there may be conflicts between them. For example, the `c8y` and `az` mappers may have entirely different external ID generation logic. So, if both published the generated `@id` value to the same entity topic, one will get overwritten. So, either they'll have to make sure that they generate non-conflicting keys or have dedicated topics for each cloud.